controller/orphan_controller.go

```diff
@@ -0,0 +1,387 @@
+	}
+}
+
+func (oc *OrphanController) processNextWorkItem() bool {
+	key, quit := oc.queue.Get()
+	if quit {
+		return false
+	}
+	defer oc.queue.Done(key)
+	err := oc.syncOrphan(key.(string))
+	oc.handleErr(err, key)
+	return true
+}
+
+func (oc *OrphanController) handleErr(err error, key interface{}) {
+	if err == nil {
+		oc.queue.Forget(key)
+		return
+	}
+
+	log := oc.logger.WithField("orphan", key)
+
+	if oc.queue.NumRequeues(key) < maxRetries {
+		log.WithError(err).Warnf("Error syncing Longhorn orphan %v: %v", key, err)
+
+		oc.queue.AddRateLimited(key)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	log.WithError(err).Warnf("Dropping Longhorn orphan %v out of the queue: %v", key, err)
+	oc.queue.Forget(key)
+}
+
+func (oc *OrphanController) syncOrphan(key string) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "%v: fail to sync orphan %v", oc.name, key)
+	}()
+
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	if namespace != oc.namespace {
+		return nil
+	}
+	return oc.reconcile(name)
+}
+
+func (oc *OrphanController) reconcile(orphanName string) (err error) {
+	orphan, err := oc.ds.GetOrphan(orphanName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		return nil
+	}
+
+	log := getLoggerForOrphan(oc.logger, orphan)
+
+	if !oc.isResponsibleFor(orphan) {
+		return nil
+	}
+
+	if orphan.Status.OwnerID != oc.controllerID {
+		orphan.Status.OwnerID = oc.controllerID
+		orphan, err = oc.ds.UpdateOrphanStatus(orphan)
+		if err != nil {
+			// we don't mind others coming first
+			if apierrors.IsConflict(errors.Cause(err)) {
+				return nil
+			}
+			return err
+		}
+		log.Infof("Orphan Controller %v picked up %v", oc.controllerID, orphan.Name)
+	}
+
+	if !orphan.DeletionTimestamp.IsZero() {
+		defer func() {
+			if err == nil {
+				err = oc.ds.RemoveFinalizerForOrphan(orphan)
+			}
+		}()
+
+		return oc.deleteOrphanedData(orphan)
+	}
+
+	existingOrphan := orphan.DeepCopy()
+	defer func() {
+		if err != nil {
+			return
+		}
+		if reflect.DeepEqual(existingOrphan.Status, orphan.Status) {
+			return
+		}
+		if _, err := oc.ds.UpdateOrphanStatus(orphan); err != nil && apierrors.IsConflict(errors.Cause(err)) {
+			log.WithError(err).Debugf("Requeue %v due to conflict", orphanName)
+			oc.enqueueOrphan(orphan)
+		}
+	}()
+
+	if err := oc.updateConditions(orphan); err != nil {
+		log.WithError(err).Errorf("failed to update conditions for orphan %v", orphan.Name)
+		return err
+	}
+
+	return nil
+}
+
+func getLoggerForOrphan(logger logrus.FieldLogger, orphan *longhorn.Orphan) *logrus.Entry {
+	return logger.WithFields(
+		logrus.Fields{
+			"orphan": orphan.Name,
+		},
+	)
+}
+
+func (oc *OrphanController) isResponsibleFor(orphan *longhorn.Orphan) bool {
+	return isControllerResponsibleFor(oc.controllerID, oc.ds, orphan.Name, orphan.Spec.NodeID, orphan.Status.OwnerID)
+}
+
+func (oc *OrphanController) deleteOrphanedData(orphan *longhorn.Orphan) (err error) {
```

Comment:
NIT: rename this to `cleanup` or `doCleanupForOrphan`. When I saw word `data`, I was thinking this function would directly delete the replica directory data.

Reply:
okay!